Documentation/ApiOverview/RequestHandling/Index.rst

```diff
@@ -0,0 +1,271 @@
+    * by acting on the request, generating the response, or forwarding the
+    * request to a subsequent middleware and possibly acting on its response.
+    */
+   interface MiddlewareInterface
+   {
+       /**
+        * Process an incoming server request
+        *
+        * Processes an incoming server request in order to produce a response.
+        * If unable to produce the response itself, it may delegate to the provided
+        * request handler to do so.
+        */
+       public function process(
+           ServerRequestInterface $request,
+           RequestHandlerInterface $handler
+       ): ResponseInterface;
+   }
+
+By doing so, the middleware can do one or multiple of the following:
+
+* Adjust the incoming request, e.g. add further information.
+
+* Create and return a PSR-7 response.
+
+* Call next request handler (which again will be a middleware).
+
+* Adjust response received from calling next request handler.
+
+.. _request-handling-middleware-examples:
+
+Middleware examples
+===================
+
+The following list shows typical use cases for middlewares.
+
+.. _request-handling-returning-custom-response:
+
+Returning custom response
+-------------------------
+
+This middleware will check whether TYPO3 is in maintenance mode and will return
+an unavailable response in that case. Otherwise the next middleware will be
+called, and their response is returned instead.
+
+::
+
+   public function process(
+       ServerRequestInterface $request,
+       RequestHandlerInterface $handler
+   ): ResponseInterface {
+       if (/* if logic */) {
+           return GeneralUtility::makeInstance(ErrorController::class)
+               ->unavailableAction(
+                   $request,
+                   'This page is temporarily unavailable.'
+               );
+       }
+
+       return $handler->handle($request);
+   }
+
+.. _request-handling-enriching-request:
+
+Enriching request
+-----------------
+
+The current request can be extended with further information, e.g. the current
+resolved site and language could be attached to the request.
+
+In order to do so, a new request is built with additional attributes, before
+calling the next request handler with enhanced request.
+
+::
+
+   public function process(
+       ServerRequestInterface $request,
+       RequestHandlerInterface $handler
+   ): ResponseInterface {
+       $routeResult = $this->matcher->matchRequest($request);
+
+       $request = $request->withAttribute('site', $routeResult->getSite());
+       $request = $request->withAttribute('language', $routeResult->getLanguage());
+
+       return $handler->handle($request);
+   }
+
+.. _request-handling-enriching-response:
+
+Enriching response
+-----------------
+
+This middleware will check the length of generated output, and add an header
+with this information to the response.
+
+In order to do so, next request handler is called. He will return the generated
```

Comment:
In order to do so, the next request handler is called. It will return the generated
response, which can be enriched before being returned.